lldb/source/Target/StackID.cpp

```diff
@@ -84,14 +84,39 @@ IsYoungerHeapCFAs(const StackID &lhs, const StackID &rhs, Process &process) {
   if (lhs_cfa_on_stack && rhs_cfa_on_stack)
     return HeapCFAComparisonResult::NoOpinion;
 
-  // FIXME: rdar://76119439
-  // At the boundary between an async parent frame calling a regular child
-  // frame, the CFA of the parent async function is a heap addresses, and the
-  // CFA of concrete child function is a stack address. Therefore, if lhs is
-  // on stack, and rhs is not, lhs is considered less than rhs, independent of
-  // address values.
+  // If one of the frames has a CFA on the stack and the other doesn't, we are
+  // at the boundary between an asynchronous and a synchronous function.
+  // Synchronous functions cannot call asynchronous functions, therefore the
+  // synchronous frame is always younger.
   if (lhs_cfa_on_stack && !rhs_cfa_on_stack)
     return HeapCFAComparisonResult::Younger;
+  if (!lhs_cfa_on_stack && rhs_cfa_on_stack)
+    return HeapCFAComparisonResult::Older;
+
+  const lldb::addr_t lhs_cfa = lhs.GetCallFrameAddress();
+  const lldb::addr_t rhs_cfa = rhs.GetCallFrameAddress();
+  // If the cfas are the same, fallback to the usual scope comparison.
+  if (lhs_cfa == rhs_cfa)
+    return HeapCFAComparisonResult::NoOpinion;
+
+  const auto ptr_size = process.GetAddressByteSize();
+
+  // Both CFAs are on the heap and they are distinct.
+  // LHS is younger if and only if its continuation async context is (directly
+  // or indirectly) RHS. Chase continuation pointers to check this case, until
+  // we hit the end of the chain (parent_ctx == 0) or a safety limit in case of
+  // an invalid continuation chain.
+  auto max_num_frames = 1024;
```

Comment:
should this limit be lower? Do we have an data to guide the choice of limit?

Reply:
In the absence of data, I'd go with the high number to protect against catastrophic failure.

Reply:
I am worried the number of memory reads could lead to catastrophic performance.

I added @jasonmolenda as a reviewer, because one of the lessons I learned from one of my previous StackID changes is that doing a `ReadMemory` to compare StackIDs can have a non-trivial impact on performance. Doing 1024 memory reads, for a single comparison, could lead to bad problems.

Reply:
In the worst case scenario, where LHS is older and therefore RHS is not going to be found on the continuation chain, we will do as many memory reads as virtual frames on the stack.

And we only do _any_ memory reads at all if we're comparing virtual frames with each other.

This is pretty much the price of doing business with async, AFAICT there are no other ways of comparing async frames with each other.

Reply:
> This is pretty much the price of doing business with async

I'm not suggesting there's a more efficient way (though there may be improvements to consider, such as caching).

I'm referring to "safety limit in case of an invalid continuation chain". Given invalid data/chain, is 1024 our best choice?

Reply:
That's fair. I don't have strong opinions here, as we should never ever hit this case. In the exceptional case where memory corruption happened and the chain is invalid, I think waiting a couple of seconds is not a big problem.

Maybe 1024 is too big of a number, but we need to be careful not to bail on "realistic" stack sizes.

Reply:
From a perf point of view, I'm less concerned about the memory read because we should only hit this codepath when we have at least one StackID that's not in stack memory, right.  We will have targets that we connect to (e.g. JTAG debuggers, maybe Linux) where we may not know if a memory region is stack or heap memory.  If we started treating StackIDs in "unknown memory type" as "could be in heap" and start dereferencing them, I'd have concerns.  This is REALLY only appropriate when we're working with swift stack frames but I don't think we've encoded that anywhere, so we have to consider unusual environments were this might get incorrectly triggered.

Reply:
>we should only hit this codepath when we have at least one StackID that's not in stack memory, right

It's slightly better than that, we only hit this codepath if _both_ StackIDs are not in stack memory.

>  If we started treating StackIDs in "unknown memory type" as "could be in heap" and start dereferencing them, I'd have concerns

I believe the current implementation of `IsCFAOnStack`  can be problematic in this respect, it doesn't answer "unknown" if `process.GetMemoryRegionInfo` fails. Instead, the default answer is "on the heap" for failures, we should probably swap that.
Curiously, the default used to be "On Stack", but then the optimization introduced in January probably swapped that unintentionally:

https://github.com/swiftlang/llvm-project/pull/7877/files#diff-f205cf833c31f7eeae9d2656bfb207b60a9f03c010c4d6ceb015d0c0048612d3R30

Note, it used to be:

```
bool StackID::IsStackAddress(...) {
 ....
 return true; // assumed
}
```

But then it became:
```
{
  m_cfa_on_stack = eLazyBoolNo;
  .... check process, may fail ....    
  return m_cfa_on_stack == eLazyBoolYes;
}
```

Reply:
> > we should only hit this codepath when we have at least one StackID that's not in stack memory, right
> 
> It's slightly better than that, we only hit this codepath if _both_ cfas are not in stack memory.

My concern is a stub that doesn't support memory region types at all, or can't distinguish stack & heap, I want to make sure every stack frame wouldn't say "is heap" and we start doing the swift specific async context chain follow.

But looking at StackID::IsCFAOnStack(),

```
bool StackID::IsCFAOnStack(Process &process) const {
  if (m_cfa_on_stack == eLazyBoolCalculate) {
    m_cfa_on_stack = eLazyBoolNo;
    if (m_cfa != LLDB_INVALID_ADDRESS) {
      MemoryRegionInfo mem_info;
      if (process.GetMemoryRegionInfo(m_cfa, mem_info).Success())
        if (mem_info.IsStackMemory() == MemoryRegionInfo::eYes)
          m_cfa_on_stack = eLazyBoolYes;
    }
  }
  return m_cfa_on_stack == eLazyBoolYes;
```


It's only going to return true when we got the memory region info and it is specifically said to be stack.  I think we're OK here.  

(the snippet you quoted earlier missed the middle bit which is important)

Reply:
Ah wait, I see the problem you're pointing out.  If we have a stub that doesn't support memory regions, or doesn't distinguish between stack & heap, this will say it's *not* on stack.  :/ We can require an affirmative "is heap" check.  I'd have to see when I added that to debugserver, and idk what we get on linux.

Reply:
e.g. if there was a StackID::IsCFAOnStack and a StackID::IsCFAOnHeap, it's possible that both of these may return false, or basically "unknown".

Reply:
Or StackID::IsCFAOnStack should return `true` unless (1) we can get the memory region, (2) the memory region is not stack, and (3) the memory region IS heap.  Require a positive match of the address in a heap memory region, or else we say it is stack because we just don't know.

Reply:
I mean, we can definitely explore more elaborate solutions, but I think for now we can just flip the conservative answer to be "on stack" always. It's never wrong to assume that (none of the upstream code has this stack vs heap distinction for StackIDs), we'll just fall back to the "regular" unwinding / stepping, which is undesirable but fine for those scenarios

Reply:
Yep, you're right, that's a simple fix that will address my worry here.